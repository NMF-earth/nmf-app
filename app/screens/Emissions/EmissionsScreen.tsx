import React from "react";
import { FlatList, SafeAreaView, View } from "react-native";
import { useSelector } from "react-redux";
import { t } from "../../utils";
import styles from "./EmissionsScreen.styles";
import {
  TabbedView,
  Text,
  EmissionListItem,
  EmissionListItemProps
} from "../../components";
import { AddEmissionAndMitigateButtons } from "./components";
import { selectors } from "./ducks";
import { navigate } from "../../navigation";

const EmissionsScreen = props => {
  const navigator = navigate(props.navigation);
  const emissionsToMitigate = useSelector(selectors.getEmissionsToMitigate);
  const emissionsMitigated = useSelector(selectors.getEmissionsMitigated);

  return (
    <React.Fragment>
      <SafeAreaView style={styles.container}>
        <TabbedView
          items={[
            {
              title: t("EMISSIONS_SCREEN_TO_OFFSET"),
              component: (
                <FlatList<EmissionListItemProps>
                  ListFooterComponent={<View style={styles.separator} />}
                  style={styles.listContainer}
                  data={emissionsToMitigate}
                  keyExtractor={item => item.id}
                  renderItem={({ item }: { item: EmissionListItemProps }) => (
                    <EmissionListItem
                      id={item.id}
                      name={item.name}
                      onPress={() =>
                        navigator.openEmissionItem({ id: item.id })
                      }
                      title={item.title}
                      co2value={item.co2value}
                      food={item.food}
                      transport={item.transport}
                      custom={item.custom}
                    />
                  )}
                />
              )
            },
            {
              title: t("EMISSIONS_SCREEN_MITIGATED"),
              component: emissionsMitigated.length ? (
                <FlatList<EmissionListItemProps>
                  ListFooterComponent={<View style={styles.separator} />}
                  style={styles.listContainer}
                  data={emissionsMitigated}
                  keyExtractor={item => item.id}
                  renderItem={({ item }: { item: EmissionListItemProps }) => (
                    <EmissionListItem
                      id={item.id}
                      name={item.name}
                      onPress={item.onPress}
                      title={item.title}
                      co2value={item.co2value}
                      food={item.food}
                      transport={item.transport}
                      custom={item.custom}
                    />
                  )}
                />
              ) : (
                <Text.Primary style={styles.textNoEmission} center lightGray>
                  {t("EMISSIONS_SCREEN_NO_EMISSION_MITIGATED")}
                </Text.Primary>
              )
            }
          ]}
        />
      </SafeAreaView>
      <AddEmissionAndMitigateButtons navigation={props.navigation} />
    </React.Fragment>
  );
};

export default EmissionsScreen;
