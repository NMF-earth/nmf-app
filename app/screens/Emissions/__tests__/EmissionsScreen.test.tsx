import React from "react";
import { create } from "react-test-renderer";

import EmissionsScreen from "../EmissionsScreen";

it("EmissionsScreen renders correctly", () => {
  const tree = create(<EmissionsScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});
