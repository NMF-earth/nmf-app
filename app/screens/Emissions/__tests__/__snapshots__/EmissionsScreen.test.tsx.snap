// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmissionsScreen renders correctly 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "co2value": 0.0803,
            "creationDate": "2020-03-01T12:08:16.623Z",
            "emissionModelType": "boat",
            "emissionType": "transport",
            "iconName": "md-boat",
            "id": "4",
            "onPress": [Function],
            "periodType": "monthly",
            "times": 1,
            "title": "Boat",
            "value": 100,
            "weekDays": Array [],
          },
        ],
        "title": "Recurring emissions",
      },
      Object {
        "data": Array [
          Object {
            "co2value": 20,
            "creationDate": "2020-03-01T12:08:16.623Z",
            "emissionModelType": "beans",
            "emissionType": "food",
            "iconName": "md-nutrition",
            "id": "2",
            "isMitigated": false,
            "onPress": [Function],
            "title": "Beans",
            "value": 10,
          },
          Object {
            "co2value": 0.0803,
            "creationDate": "2020-03-01T12:08:16.623Z",
            "emissionModelType": "boat",
            "emissionType": "transport",
            "iconName": "md-boat",
            "id": "3",
            "isMitigated": true,
            "onPress": [Function],
            "title": "Boat",
            "value": 100,
          },
        ],
        "date": "2020-03-01T00:00:00+00:00",
      },
      Object {
        "data": Array [
          Object {
            "co2value": 60,
            "creationDate": "2020-01-01T12:08:16.623Z",
            "emissionModelType": "beans",
            "emissionType": "food",
            "iconName": "md-nutrition",
            "id": "1",
            "isMitigated": false,
            "onPress": [Function],
            "title": "Beans",
            "value": 30,
          },
        ],
        "date": "2020-01-01T00:00:00+00:00",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={
    Array [
      0,
      3,
      7,
    ]
  }
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <SectionHeader
        title="Recurring emissions"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionListItem
        co2value={0.0803}
        emissionModelType="boat"
        iconName="md-boat"
        id="4"
        onPress={[Function]}
        title="Boat"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
      style={null}
    >
      <SectionHeader
        date="2020-03-01T00:00:00+00:00"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionListItem
        co2value={20}
        emissionModelType="beans"
        iconName="md-nutrition"
        id="2"
        isMitigated={false}
        onPress={[Function]}
        title="Beans"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionListItem
        co2value={0.0803}
        emissionModelType="boat"
        iconName="md-boat"
        id="3"
        isMitigated={true}
        onPress={[Function]}
        title="Boat"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
      style={null}
    >
      <SectionHeader
        date="2020-01-01T00:00:00+00:00"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionListItem
        co2value={60}
        emissionModelType="beans"
        iconName="md-nutrition"
        id="1"
        isMitigated={false}
        onPress={[Function]}
        title="Beans"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "height": 100,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;

exports[`EmissionsScreen renders correctly if no emissions 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "data": Array [],
        "title": null,
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={
    Array [
      0,
    ]
  }
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <SectionHeader
        title={null}
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "height": 100,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;

exports[`EmissionsScreen renders correctly if only normal emissions 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "co2value": 20,
            "creationDate": "2020-03-01T12:08:16.623Z",
            "emissionModelType": "beans",
            "emissionType": "food",
            "iconName": "md-nutrition",
            "id": "2",
            "isMitigated": false,
            "onPress": [Function],
            "title": "Beans",
            "value": 10,
          },
          Object {
            "co2value": 0.0803,
            "creationDate": "2020-03-01T12:08:16.623Z",
            "emissionModelType": "boat",
            "emissionType": "transport",
            "iconName": "md-boat",
            "id": "3",
            "isMitigated": true,
            "onPress": [Function],
            "title": "Boat",
            "value": 100,
          },
        ],
        "date": "2020-03-01T00:00:00+00:00",
      },
      Object {
        "data": Array [
          Object {
            "co2value": 60,
            "creationDate": "2020-01-01T12:08:16.623Z",
            "emissionModelType": "beans",
            "emissionType": "food",
            "iconName": "md-nutrition",
            "id": "1",
            "isMitigated": false,
            "onPress": [Function],
            "title": "Beans",
            "value": 30,
          },
        ],
        "date": "2020-01-01T00:00:00+00:00",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={
    Array [
      0,
      4,
    ]
  }
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <SectionHeader
        date="2020-03-01T00:00:00+00:00"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionListItem
        co2value={20}
        emissionModelType="beans"
        iconName="md-nutrition"
        id="2"
        isMitigated={false}
        onPress={[Function]}
        title="Beans"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionListItem
        co2value={0.0803}
        emissionModelType="boat"
        iconName="md-boat"
        id="3"
        isMitigated={true}
        onPress={[Function]}
        title="Boat"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
      style={null}
    >
      <SectionHeader
        date="2020-01-01T00:00:00+00:00"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionListItem
        co2value={60}
        emissionModelType="beans"
        iconName="md-nutrition"
        id="1"
        isMitigated={false}
        onPress={[Function]}
        title="Beans"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "height": 100,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;

exports[`EmissionsScreen renders correctly if only recurring emissions 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "co2value": 0.0803,
            "creationDate": "2020-03-01T12:08:16.623Z",
            "emissionModelType": "boat",
            "emissionType": "transport",
            "iconName": "md-boat",
            "id": "4",
            "onPress": [Function],
            "periodType": "monthly",
            "times": 1,
            "title": "Boat",
            "value": 100,
            "weekDays": Array [],
          },
        ],
        "title": "Recurring emissions",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={
    Array [
      0,
    ]
  }
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <SectionHeader
        title="Recurring emissions"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <EmissionListItem
        co2value={0.0803}
        emissionModelType="boat"
        iconName="md-boat"
        id="4"
        onPress={[Function]}
        title="Boat"
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "height": 100,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;
